### DISCUSSION TITLE
AttributeError: 'dict' object has no attribute 'geometry

### DISCUSSION THREAD

**Discussion by AdrianKriger** [cityjson/cjio#158] [Q&A] [answered]
Happy New Year.

With:

```python
cm = [cityjson.load(path=file]]
rds = cm.get_cityobjects(id='1000004')
rds.geometry

Traceback (most recent call last):

  Cell In[131], line 5
    rds.geometry

AttributeError: 'dict' object has no attribute 'geometry'
```

When I parse the city model through [val3dity](http://geovalidation.bk.tudelft.nl/val3dity/); it returns a [102 CONSECUTIVE_POINTS_SAME](https://val3dity.readthedocs.io/en/latest/errors/#consecutive-points-same) error for a few road objects. 
I want to see these repeated points. 
How can I see the vertices please?

file -> [here](https://github.com/AdrianKriger/osm_LoD1_3DCityModel/blob/main/village_campus/extra/citjsnClean_test3d-UE-str_cpsame.json)

**Comment by hugoledoux** [comment, marked as answer]
`get_cityobjects()` returns all the objects, so you need to specify which one (even if only one)

```python
cm = [cityjson.load(path=file]]
rds = cm.get_cityobjects(id='1000004')
rds['1000004'].geometry
```

**Reply by AdrianKriger** [reply]
As usual your level of service is fantastic.
Thank you.

```python
geom = rds['1000004'].geometry[0]
geom.boundaries
```
Gives:
```python
 ...[[[264596.532, 6241575.696, 101.45],
   [264596.13, 6241575.648, 101.45],
   [264596.265, 6241575.657, 101.45]]],
 [[[264595.656, 6241575.666, 101.45],
   [264595.726, 6241575.66, 101.45],
   [264595.593, 6241575.677, 101.45]]],
 [[[264596.265, 6241575.657, 101.45],
   [264596.399, 6241575.673, 101.45],
   [264596.532, 6241575.696, 101.45]]],
 [[[264596.265, 6241575.657, 101.45],
   [264596.336, 6241575.662, 101.45],
   [264596.399, 6241575.673, 101.45]]],
 [[[264595.726, 6241575.66, 101.45],
   [264596.532, 6241575.696, 101.45],
   [264595.593, 6241575.677, 101.45]]],
 [[[264595.529, 6241575.685, 101.45],
   [264595.743, 6241581.233, 101.45],
   [264593.295, 6241576.081, 101.45]]],...
```
Some of the points are close (these are possible cuvilinear road features) but I don't see duplicates. 
Could a tolerance value be producing the error? 
Which validation option would I change? _--- I'm in a bit restricted with computer / processing or I would run val3dity locally_. 

How common is this type of error? 

The city model has gone from [305 MULTIPLE_CONNECTED_COMPONENTS](https://val3dity.readthedocs.io/en/latest/errors/#multiple-connected-components) to [102 CONSECUTIVE_POINTS_SAME](https://val3dity.readthedocs.io/en/latest/errors/#consecutive-points-same). 
Is this an improvement?

**Reply by hugoledoux** [reply]
> Could a tolerance value be producing the error?

yes, val3dity has a snap tolerance (https://val3dity.readthedocs.io/en/latest/usage/#snap-tol), you can play with it.

Dunno if improvements, those 2 could be related.

**Reply by AdrianKriger** [reply]
Ok. 
snap tolerance=0.0001 results in 1 invalid 3D geometry -> TINRelief: 305	[MULTIPLE_CONNECTED_COMPONENTS](https://val3dity.readthedocs.io/en/latest/errors/#multiple-connected-components).

This is what I had before. 
0.0001 seems a bit excessive. If I understand this correctly its saying; _some polygons features are 0.0001 meters from the TINRelief_.
Is this correct? I can live with a fraction of a mm. 
What do downstream applications need? Will this become a challenge if the model is used in later CFD, etc. applications?

**Reply by hugoledoux** [reply]
no not correct. This says: “if 2 points are closer then X, then we assume that they are the same”. Those points need to be part of the same City Objects, the validation is PER OBJECT, nothing between objects is performed by val3dity.

**Reply by AdrianKriger** [reply]
Let me see if I understand:
snap_tolerance applies to the [102 CONSECUTIVE_POINTS_SAME](https://val3dity.readthedocs.io/en/latest/errors/#consecutive-points-same) error. With a snap_tolerance of 0.0001 we're saying: points, where the mm are different, are not the same.
[305 MULTIPLE_CONNECTED_COMPONENTS](https://val3dity.readthedocs.io/en/latest/errors/#multiple-connected-components) is concerned with the _'islands_' in the TINRelief (ground area in the holes [courtyards] of buildings seperated from the larger terrain).

How would I prevent the [305](https://val3dity.readthedocs.io/en/latest/errors/#multiple-connected-components)? Would cjio.triangulate() help?
